editor/gui/scene_tree_editor.cpp

```diff
@@ -599,45 +704,134 @@ void SceneTreeEditor::_set_item_custom_color(TreeItem *p_item, Color p_color) {
 }
 
 void SceneTreeEditor::_node_script_changed(Node *p_node) {
-	if (tree_dirty) {
+	HashMap<Node *, CachedNode>::Iterator I = node_cache.get(p_node);
+	if (!I) {
+		// We leave these signals connected when switching tabs.
+		// If the node is not in cache it was for a different tab.
 		return;
 	}
 
-	callable_mp(this, &SceneTreeEditor::_update_tree).call_deferred(false);
-	tree_dirty = true;
+	node_cache.mark_dirty(p_node);
+
+	_update_if_clean();
 }
 
-void SceneTreeEditor::_node_removed(Node *p_node) {
-	if (EditorNode::get_singleton()->is_exiting()) {
-		return; //speed up exit
+void SceneTreeEditor::_move_node_children(HashMap<Node *, CachedNode>::Iterator &p_I) {
+	TreeItem *item = p_I->value.item;
+	Node *node = p_I->key;
+	int cc = node->get_child_count(false);
+
+	for (int i = 0; i < cc; i++) {
+		HashMap<Node *, CachedNode>::Iterator CI = node_cache.get(node->get_child(i, false));
+		if (CI) {
+			_move_node_item(item, CI);
+		}
+	}
+
+	p_I->value.has_moved_children = false;
+}
+
+void SceneTreeEditor::_move_node_item(TreeItem *p_parent, HashMap<Node *, CachedNode>::Iterator &p_I) {
+	if (!p_parent) {
+		return;
 	}
 
-	if (p_node->is_connected(CoreStringName(script_changed), callable_mp(this, &SceneTreeEditor::_node_script_changed))) {
-		p_node->disconnect(CoreStringName(script_changed), callable_mp(this, &SceneTreeEditor::_node_script_changed));
+	Node *node = p_I->key;
+
+	int current_node_index = node->get_index(false);
+	TreeItem *item = p_I->value.item;
+
+	if (item->get_parent() != p_parent) {
+		TreeItem *p = item->get_parent();
+		if (p) {
+			item->get_parent()->remove_child(item);
+		}
+		p_parent->add_child(item);
+		p_I->value.index = -1;
```

Comment:
I didn't do that on purpose as this might uselessly build the child cache if we don't need it, which is more expensive than a single TreeItem::get_index()

Reply:
`TreeItem::get_child()` also calls `_create_children_cache()` to rebuild the cache. Rebuilding the cache is unavoidable in most cases.

This code snippet will rebuild the cache when moving to a new parent node with no children. Otherwise, the cache may already exist.

Reply:
`Tree::draw_item()` calls `Tree::get_visible_child_count()`, which usually rebuilds the cache. So the expanded item usually has the cache rebuilt.

https://github.com/godotengine/godot/blob/aa8d9b83f66488dbb2c9c918e9016ef80f821fb4/scene/gui/tree.cpp#L2615-L2617

If the cache is already created, using `p_parent->get_child_count()` to determine the index is cheaper than using `item->get_index()`.

Reply:
But in the "happy case" we don't call TreeItem::get_child() at all. The happy case is:

* Add the node to parent (TreeItem::add_child() does not create a cache)
* Node is already in the right place (TreeItem::get_index() is cheaper than building the index)

I understand what you are saying, but I don't think that adding the cost of building the cache to the "happy path" is the right choice.

Reply:
In most cases, we need to see the item before we can operate it. In order to see it, `Tree::draw_item()` is called, which creates the cache. So the cost of creating the cache is not here.

Reply:
I implemented it a little bit differently, but this also avoids the TreeItem::get_index() when it is not required.

Thank you!